PR: https://github.com/odoo/odoo/pull/180813

From: 01b87f1230beac0568f4e3b1b76e547909506892
From: william-andre
Date: 2024-09-21 13:34:59

Breaking data model changes scores: del:10 + add:4, change matches:
-    payment_id = fields.Many2one(
+    matched_payment_ids = fields.Many2many(
-    _inherits = {'account.move': 'move_id'}
-    is_internal_transfer = fields.Boolean(string="Internal Transfer",
-    destination_journal_id = fields.Many2one(
+    invoice_ids = fields.Many2many(
-    duplicate_move_ids = fields.Many2many(comodel_name='account.move', compute='_compute_duplicate_move_ids')
+    duplicate_payment_ids = fields.Many2many(comodel_name='account.payment', compute='_compute_duplicate_payment_ids')
+    attachment_ids = fields.One2many('ir.attachment', 'res_id', string='Attachments')
-    account_journal_payment_debit_account_id = fields.Many2one('account.account', string='Journal Outstanding Receipts', check_company=True)
-    account_journal_payment_credit_account_id = fields.Many2one('account.account', string='Journal Outstanding Payments', check_company=True)
-    account_journal_payment_debit_account_id = fields.Many2one(
-    account_journal_payment_credit_account_id = fields.Many2one(

Total Changes: 831

[IMP] account: make journal entry optional for payments

TLDR
====

A payment doesn't always generate a journal entry anymore, it will only
do so if an outstanding account is set.
This means that the fallback on the company for outstanding
receipts/payment has been removed as well.

Purpose
=======

Payment management in Odoo should be treated as an extra-accounting
process and should not generate accounting entries automatically.
Recording payments is primarily used to prevent duplicate payments of
the same invoice and to prepare bank files, such as SEPA Credit
Transfers. However, it is common for payments to be canceled, rejected,
or deleted due to errors, such as insufficient funds in the client’s
account or forgotten file submissions.

Another goal of this simplification is to make bank reconciliation
easier. The process involves simply identifying the partner and, if
possible, matching a document. If immediate matching isn’t feasible,
it’s not a problem; you can associate the bank transaction with the
partner’s account and complete the matching later. This approach
minimizes errors and greatly simplifies the workflow. Currently, the
procedure is more complex, often requiring matching within the
outstanding account or directly with the partner, which increases
confusion and the risk of mistakes (in cases of payments).

Currently, the accounting entries generated by payments can clutter the
customer and vendor balances, making data interpretation more
challenging. For example, instead of seeing only the invoice entries
related to bank transactions, users often see payment entries, which
complicates management.

Dates are also an issue, as we often see the payment date, but
ultimately, only the bank date is truly useful.

Implementation
==============

Before the payment used to be a composition based inheritance (a.k.a.
`_inherits`) of the journal entry, but the journal entry doesn't always
exist anymore, we can't rely on that mechanism anymore.
* Payments can be generated from scratch (the "New" button), in that
  case, the payment will generate it's journal entry upon validation, if
  and only is there is an outstanding account set.
* Payments can also be generated from various other sources (i.e.
  "Register Payment", from Expenses, from the PoS,...). In those cases,
  the source is responsible to generate the journal entry and link it to
  the payment. By simplicity, the lines to be created can be provided to
  the create values. In these case the payment won't go through the draft
  state.
* This also means that there is no synchronisation anymore between the
  payment and the journal entry.
* Because there isn't always a journal entry anymore, we needed to add
  some fields on the payment which might be duplidated on the journal
  entry if there is one: Sent, Memo, Name and State.
* Now that the payments can not have a journal entry, it is impossible to
  rely only on the reconciliation mechanism to link the payments and the
  invoices. New many2many fields `invoice_ids`/`matched_payment_ids` have
  been added between the two models.

Details
=======

* There is now a "(Un)Block Payment" button on the invoice in order to
  prevent registering a payment if needed.
* When an invoice is fully reconciled (with a bank transaction), the
  related payments will be marked as Paid. More and improved heuristics
  will be added in the future to avoid keeping pending payments as In
  Process.
* Payments without journal entries don't appear in the paid amount, but
  information about pending payments will be added in the future on
  invoice reports and the followup.
* New states have been added on the payment. Posted is renamed in To
  Process, and 3 completely new states are added:
    * Paid: there was previously a checkbox Matched, but that was only
      relevant for reconciliation, which is not possible anymore
    * Canceled: the user doesn't want to proceed with this payment
      anymore
    * Rejected: the payment was rejected by the payment provider
      (insufficient funds, wrong details,...)
* Internal Transfers have been removed: it should be matched as a
  writeoff during bank reconciliation instead.

task-4180163

closes odoo/odoo#180813

Related: odoo/enterprise#70349
Related: odoo/upgrade#6539
Signed-off-by: Quentin De Paoli (qdp) <qdp@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/account_account.py
+++ b/addons/account/models/account_account.py
@@ -226,7 +226,7 @@ class AccountAccount(models.Model):
             JOIN res_company company ON company.id = journal.company_id
             JOIN account_payment_method_line apml ON apml.journal_id = journal.id
             JOIN account_payment_method apm on apm.id = apml.payment_method_id
-            JOIN account_account account ON account.id = COALESCE(apml.payment_account_id, company.account_journal_payment_debit_account_id)
+            JOIN account_account account ON account.id = apml.payment_account_id
             WHERE journal.currency_id IS NOT NULL
             AND journal.currency_id != company.currency_id
             AND account.currency_id != journal.currency_id
@@ -242,7 +242,7 @@ class AccountAccount(models.Model):
             JOIN res_company company ON company.id = journal.company_id
             JOIN account_payment_method_line apml ON apml.journal_id = journal.id
             JOIN account_payment_method apm on apm.id = apml.payment_method_id
-            JOIN account_account account ON account.id = COALESCE(apml.payment_account_id, company.account_journal_payment_credit_account_id)
+            JOIN account_account account ON account.id = apml.payment_account_id
             WHERE journal.currency_id IS NOT NULL
             AND journal.currency_id != company.currency_id
             AND account.currency_id != journal.currency_id
@@ -302,7 +302,6 @@ class AccountAccount(models.Model):
             return
 
         self.env['account.journal'].flush_model(['company_id', 'default_account_id'])
-        self.env['res.company'].flush_model(['account_journal_payment_credit_account_id', 'account_journal_payment_debit_account_id'])
         self.env['account.payment.method.line'].flush_model(['journal_id', 'payment_account_id'])
 
         self._cr.execute('''
@@ -311,12 +310,6 @@ class AccountAccount(models.Model):
             JOIN res_company company on journal.company_id = company.id
             LEFT JOIN account_payment_method_line apml ON journal.id = apml.journal_id
             WHERE (
-                company.account_journal_payment_credit_account_id IN %(accounts)s
-                AND company.account_journal_payment_credit_account_id != journal.default_account_id
-                ) OR (
-                company.account_journal_payment_debit_account_id in %(accounts)s
-                AND company.account_journal_payment_debit_account_id != journal.default_account_id
-                ) OR (
                 apml.payment_account_id IN %(accounts)s
                 AND apml.payment_account_id != journal.default_account_id
             )

--- a/addons/account/models/account_journal.py
+++ b/addons/account/models/account_journal.py
@@ -1036,7 +1036,7 @@ class AccountJournal(models.Model):
         self.ensure_one()
         account_ids = set()
         for line in self.inbound_payment_method_line_ids:
-            account_ids.add(line.payment_account_id.id or self.company_id.account_journal_payment_debit_account_id.id)
+            account_ids.add(line.payment_account_id.id)
         return self.env['account.account'].browse(account_ids)
 
     def _get_journal_outbound_outstanding_payment_accounts(self):
@@ -1046,7 +1046,7 @@ class AccountJournal(models.Model):
         self.ensure_one()
         account_ids = set()
         for line in self.outbound_payment_method_line_ids:
-            account_ids.add(line.payment_account_id.id or self.company_id.account_journal_payment_credit_account_id.id)
+            account_ids.add(line.payment_account_id.id)
         return self.env['account.account'].browse(account_ids)
 
     def _get_available_payment_method_lines(self, payment_type):

--- a/addons/account/models/account_journal_dashboard.py
+++ b/addons/account/models/account_journal_dashboard.py
@@ -810,7 +810,7 @@ class account_journal(models.Model):
                    END) AS amount_total,
                    SUM(amount_company_currency_signed) AS amount_total_company
               FROM account_payment payment
-              JOIN account_move move ON move.payment_id = payment.id
+              JOIN account_move move ON move.origin_payment_id = payment.id
               JOIN account_journal journal ON move.journal_id = journal.id
              WHERE payment.is_matched IS TRUE
                AND move.state = 'posted'
@@ -838,7 +838,7 @@ class account_journal(models.Model):
                    END) AS amount_total,
                    SUM(amount_company_currency_signed) AS amount_total_company
               FROM account_payment payment
-              JOIN account_move move ON move.payment_id = payment.id
+              JOIN account_move move ON move.origin_payment_id = payment.id
              WHERE (NOT payment.is_matched OR payment.is_matched IS NULL)
                AND move.state = 'posted'
                AND payment.journal_id = ANY(%s)

--- a/addons/account/models/account_move.py
+++ b/addons/account/models/account_move.py
@@ -50,6 +50,7 @@ PAYMENT_STATE_SELECTION = [
         ('paid', 'Paid'),
         ('partial', 'Partially Paid'),
         ('reversed', 'Reversed'),
+        ('blocked', 'Blocked'),
         ('invoicing_legacy', 'Invoicing App Legacy'),
 ]
 
@@ -193,13 +194,21 @@ class AccountMove(models.Model):
     )
 
     # === Payment fields === #
-    payment_id = fields.Many2one(
+    origin_payment_id = fields.Many2one(  # the payment this is the journal entry of
         comodel_name='account.payment',
         string="Payment",
         index='btree_not_null',
         copy=False,
         check_company=True,
     )
+    matched_payment_ids = fields.Many2many(  # the payments linked to this invoice
+        string="Matched Payments",
+        comodel_name='account.payment',
+        relation='account_move__account_payment',
+        column1='invoice_id',
+        column2='payment_id',
+    )
+    payment_count = fields.Integer(compute='_compute_payment_count')
 
     # === Statement fields === #
     statement_line_id = fields.Many2one(
@@ -787,7 +796,7 @@ class AccountMove(models.Model):
             if move.journal_id.company_id not in move.company_id.parent_ids:
                 move.company_id = (move.journal_id.company_id or self.env.company)._accessible_branches()[:1]
 
-    @api.depends('move_type', 'payment_id', 'statement_line_id')
+    @api.depends('move_type', 'origin_payment_id', 'statement_line_id')
     def _compute_journal_id(self):
         for move in self.filtered(lambda r: r.journal_id.type not in r._get_valid_journal_types()):
             move.journal_id = move._search_default_journal()
@@ -797,7 +806,7 @@ class AccountMove(models.Model):
             return ['sale']
         elif self.is_purchase_document(include_receipts=True):
             return ['purchase']
-        elif self.payment_id or self.statement_line_id or self.env.context.get('is_payment') or self.env.context.get('is_statement_line'):
+        elif self.origin_payment_id or self.statement_line_id or self.env.context.get('is_payment') or self.env.context.get('is_statement_line'):
             return ['bank', 'cash', 'credit']
         return ['general']
 
@@ -845,7 +854,7 @@ class AccountMove(models.Model):
                 ('type', '=', journal_type),
             ])
 
-    @api.depends('posted_before', 'state', 'journal_id', 'date', 'move_type', 'payment_id')
+    @api.depends('posted_before', 'state', 'journal_id', 'date', 'move_type', 'origin_payment_id')
     def _compute_name(self):
         self = self.sorted(lambda m: (m.date, m.ref or '', m._origin.id))
 
@@ -1021,10 +1030,10 @@ class AccountMove(models.Model):
                 invoice.direction_sign = -1
 
     @api.depends(
-        'line_ids.matched_debit_ids.debit_move_id.move_id.payment_id.is_matched',
+        'line_ids.matched_debit_ids.debit_move_id.move_id.origin_payment_id.is_matched',
         'line_ids.matched_debit_ids.debit_move_id.move_id.line_ids.amount_residual',
         'line_ids.matched_debit_ids.debit_move_id.move_id.line_ids.amount_residual_currency',
-        'line_ids.matched_credit_ids.credit_move_id.move_id.payment_id.is_matched',
+        'line_ids.matched_credit_ids.credit_move_id.move_id.origin_payment_id.is_matched',
         'line_ids.matched_credit_ids.credit_move_id.move_id.line_ids.amount_residual',
         'line_ids.matched_credit_ids.credit_move_id.move_id.line_ids.amount_residual_currency',
         'line_ids.balance',
@@ -1079,7 +1088,7 @@ class AccountMove(models.Model):
             move.amount_residual_signed = total_residual
             move.amount_total_in_currency_signed = abs(move.amount_total) if move.move_type == 'entry' else -(sign * move.amount_total)
 
-    @api.depends('amount_residual', 'move_type', 'state', 'company_id')
+    @api.depends('amount_residual', 'move_type', 'state', 'company_id', 'matched_payment_ids')
     def _compute_payment_state(self):
         stored_ids = tuple(self.ids)
         if stored_ids:
@@ -1098,7 +1107,7 @@ class AccountMove(models.Model):
                         account.account_type AS source_line_account_type,
                         ARRAY_AGG(counterpart_move.move_type) AS counterpart_move_types,
                         COALESCE(BOOL_AND(COALESCE(pay.is_matched, FALSE))
-                            FILTER (WHERE counterpart_move.payment_id IS NOT NULL), TRUE) AS all_payments_matched,
+                            FILTER (WHERE counterpart_move.origin_payment_id IS NOT NULL), TRUE) AS all_payments_matched,
                         BOOL_OR(COALESCE(BOOL(pay.id), FALSE)) as has_payment,
                         BOOL_OR(COALESCE(BOOL(counterpart_move.statement_line_id), FALSE)) as has_st_line
                     FROM account_partial_reconcile part
@@ -1106,7 +1115,7 @@ class AccountMove(models.Model):
                     JOIN account_account account ON account.id = source_line.account_id
                     JOIN account_move_line counterpart_line ON counterpart_line.id = part.%s
                     JOIN account_move counterpart_move ON counterpart_move.id = counterpart_line.move_id
-                    LEFT JOIN account_payment pay ON pay.id = counterpart_move.payment_id
+                    LEFT JOIN account_payment pay ON pay.id = counterpart_move.origin_payment_id
                     WHERE source_line.move_id IN %s AND counterpart_line.move_id != source_line.move_id
                     GROUP BY source_line.id, source_line.move_id, account.account_type
                 ''', SQL.identifier(source_field), SQL.identifier(counterpart_field), stored_ids))
@@ -1134,7 +1143,7 @@ class AccountMove(models.Model):
             if payment_state_matters:
                 reconciliation_vals = [x for x in reconciliation_vals if x['source_line_account_type'] in ('asset_receivable', 'liability_payable')]
 
-            new_pmt_state = 'not_paid'
+            new_pmt_state = 'not_paid' if invoice.payment_state != 'blocked' else 'blocked'
             if invoice.state == 'posted':
 
                 # Posted invoice/expense entry.
@@ -1165,10 +1174,10 @@ class AccountMove(models.Model):
                                             and reverse_move_types == {'entry'})
                             if in_reverse or out_reverse or misc_reverse:
                                 new_pmt_state = 'reversed'
-
+                    elif invoice.matched_payment_ids.filtered(lambda p: not p.move_id and p.state == 'in_process'):
+                        new_pmt_state = invoice._get_invoice_in_payment_state()
                     elif reconciliation_vals:
                         new_pmt_state = 'partial'
-
             invoice.payment_state = new_pmt_state
 
     @api.depends('payment_state', 'state')
@@ -1176,6 +1185,11 @@ class AccountMove(models.Model):
         for move in self:
             move.status_in_payment = move.state if move.state in ('draft', 'cancel') else move.payment_state
 
+    @api.depends('matched_payment_ids')
+    def _compute_payment_count(self):
+        for invoice in self:
+            invoice.payment_count = len(invoice.matched_payment_ids)
+
     @api.depends('invoice_payment_term_id', 'invoice_date', 'currency_id', 'amount_total_in_currency_signed', 'invoice_date_due')
     def _compute_needed_terms(self):
         for invoice in self.with_context(bin_size=False):
@@ -2399,7 +2413,6 @@ class AccountMove(models.Model):
             return
 
         self_sudo = self.sudo()
-        self_sudo.payment_id._synchronize_from_moves(changed_fields)
         self_sudo.statement_line_id._synchronize_from_moves(changed_fields)
 
     # -------------------------------------------------------------------------
@@ -3150,7 +3163,7 @@ class AccountMove(models.Model):
             return "WHERE FALSE", {}
         where_string = "WHERE journal_id = %(journal_id)s AND name != '/'"
         param = {'journal_id': self.journal_id.id}
-        is_payment = self.payment_id or self.env.context.get('is_payment')
+        is_payment = self.origin_payment_id or self.env.context.get('is_payment')
 
         if not relaxed:
             domain = [('journal_id', '=', self.journal_id.id), ('id', '!=', self.id or self._origin.id), ('name', 'not in', ('/', '', False))]
@@ -3158,7 +3171,7 @@ class AccountMove(models.Model):
                 refund_types = ('out_refund', 'in_refund')
                 domain += [('move_type', 'in' if self.move_type in refund_types else 'not in', refund_types)]
             if self.journal_id.payment_sequence:
-                domain += [('payment_id', '!=' if is_payment else '=', False)]
+                domain += [('origin_payment_id', '!=' if is_payment else '=', False)]
             reference_move_name = self.sudo().search(domain + [('date', '<=', self.date)], order='date desc', limit=1).name
             if not reference_move_name:
                 reference_move_name = self.sudo().search(domain, order='date asc', limit=1).name
@@ -3196,9 +3209,9 @@ class AccountMove(models.Model):
                 where_string += " AND move_type NOT IN ('out_refund', 'in_refund') "
         elif self.journal_id.payment_sequence:
             if is_payment:
-                where_string += " AND payment_id IS NOT NULL "
+                where_string += " AND origin_payment_id IS NOT NULL "
             else:
-                where_string += " AND payment_id IS NULL "
+                where_string += " AND origin_payment_id IS NULL "
 
         return where_string, param
 
@@ -3225,7 +3238,7 @@ class AccountMove(models.Model):
             starting_sequence = "%s/%s/%02d/0000" % (self.journal_id.code, year_part, self.date.month)
         if self.journal_id.refund_sequence and self.move_type in ('out_refund', 'in_refund'):
             starting_sequence = "R" + starting_sequence
-        if self.journal_id.payment_sequence and self.payment_id or self.env.context.get('is_payment'):
+        if self.journal_id.payment_sequence and self.origin_payment_id or self.env.context.get('is_payment'):
             starting_sequence = "P" + starting_sequence
         return starting_sequence
 
@@ -4258,7 +4271,7 @@ class AccountMove(models.Model):
 
     def _get_reconciled_payments(self):
         """Helper used to retrieve the reconciled payments on this journal entry"""
-        return self._get_reconciled_amls().move_id.payment_id
+        return self._get_reconciled_amls().move_id.origin_payment_id
 
     def _get_reconciled_statement_lines(self):
         """Helper used to retrieve the reconciled statement lines on this journal entry"""
@@ -4686,15 +4699,18 @@ class AccountMove(models.Model):
     # PUBLIC ACTIONS
     # -------------------------------------------------------------------------
 
+    def open_payments(self):
+        return self.matched_payment_ids._get_records_action(name=_("Payments"))
+
     def open_reconcile_view(self):
         return self.line_ids.open_reconcile_view()
 
     def action_open_business_doc(self):
         self.ensure_one()
-        if self.payment_id:
+        if self.origin_payment_id:
             name = _("Payment")
             res_model = 'account.payment'
-            res_id = self.payment_id.id
+            res_id = self.origin_payment_id.id
         elif self.statement_line_id:
             name = _("Bank Transaction")
             res_model = 'account.bank.statement.line'
@@ -4759,7 +4775,7 @@ class AccountMove(models.Model):
         return self.action_force_register_payment()
 
     def action_force_register_payment(self):
-        if any(m.payment_state not in ('not_paid', 'partial') for m in self):
+        if any(m.payment_state not in ('not_paid', 'partial', 'in_payment') for m in self):
             raise UserError(_("You can only register payments for (partially) unpaid documents."))
         if any(m.move_type == 'entry' for m in self):
             raise UserError(_("You cannot register payments for miscellaneous entries."))
@@ -4830,17 +4846,13 @@ class AccountMove(models.Model):
         return action
 
     def action_post(self):
-        moves_with_payments = self.filtered('payment_id')
-        if moves_with_payments:
-            moves_with_payments.payment_id.action_post()
-        other_moves = self - moves_with_payments
         # Disabled by default to avoid breaking automated action flow
         if (
             not self.env.context.get('disable_abnormal_invoice_detection', True)
-            and other_moves.filtered(lambda m: m.abnormal_amount_warning or m.abnormal_date_warning)
+            and self.filtered(lambda m: m.abnormal_amount_warning or m.abnormal_date_warning)
         ):
             wizard = self.env['validate.account.move'].create({
-                'move_ids': [Command.set(other_moves.ids)],
+                'move_ids': [Command.set(self.ids)],
             })
             return {
                 'name': _("Confirm Entries"),
@@ -4850,9 +4862,9 @@ class AccountMove(models.Model):
                 'view_mode': 'form',
                 'target': 'new',
             }
-        if other_moves:
-            other_moves._post(soft=False)
-        if autopost_bills_wizard := other_moves._show_autopost_bills_wizard():
+        if self:
+            self._post(soft=False)
+        if autopost_bills_wizard := self._show_autopost_bills_wizard():
             return autopost_bills_wizard
         return False
 
@@ -4950,6 +4962,16 @@ class AccountMove(models.Model):
 
         self.write({'auto_post': 'no', 'state': 'cancel'})
 
+    def action_toggle_block_payment(self):
+        self.ensure_one()
+        if self.payment_state == 'blocked':
+            self.payment_state = 'not_paid'
+            self.env.add_to_compute(self._fields['payment_state'], self)
+        else:
+            if self.payment_state in ('paid', 'in_payment'):
+                raise UserError(_("You can't block a paid invoice."))
+            self.payment_state = 'blocked'
+
     def action_activate_currency(self):
         self.currency_id.filtered(lambda currency: not currency.active).write({'active': True})
 
@@ -5701,8 +5723,8 @@ class AccountMove(models.Model):
         self.ensure_one()
 
         if not self.is_invoice(include_receipts=True):
-            if self.payment_id and 'state' in init_values:
-                self.payment_id._message_track(['state'], {self.payment_id.id: init_values})
+            if self.origin_payment_id and 'state' in init_values:
+                self.origin_payment_id._message_track(['state'], {self.origin_payment_id.id: init_values})
             return super()._track_subtype(init_values)
 
         if 'payment_state' in init_values and self.payment_state == 'paid':

--- a/addons/account/models/account_move_line.py
+++ b/addons/account/models/account_move_line.py
@@ -164,7 +164,7 @@ class AccountMoveLine(models.Model):
     payment_id = fields.Many2one(
         comodel_name='account.payment',
         string="Originator Payment",
-        related='move_id.payment_id', store=True,
+        related='move_id.origin_payment_id', store=True,
         auto_join=True,
         index='btree_not_null',
         help="The payment that created this entry")
@@ -1883,7 +1883,7 @@ class AccountMoveLine(models.Model):
         """
 
         def is_payment(aml):
-            return aml.move_id.payment_id or aml.move_id.statement_line_id
+            return aml.move_id.origin_payment_id or aml.move_id.statement_line_id
 
         def get_odoo_rate(aml, other_aml, currency):
             if forced_rate := self._context.get('forced_rate_from_register_payment'):

--- a/addons/account/models/account_payment.py
+++ b/addons/account/models/account_payment.py
@@ -8,36 +8,51 @@ from odoo.tools import SQL
 
 class AccountPayment(models.Model):
     _name = "account.payment"
-    _inherits = {'account.move': 'move_id'}
     _inherit = ['mail.thread.main.attachment', 'mail.activity.mixin']
     _description = "Payments"
     _order = "date desc, name desc"
     _check_company_auto = True
 
     # == Business fields ==
+    name = fields.Char(string="Number", compute='_compute_name', store=True)
+    date = fields.Date(default=fields.Date.context_today, required=True, tracking=True)
     move_id = fields.Many2one(
         comodel_name='account.move',
-        string='Journal Entry', required=True, readonly=True, ondelete='cascade',
+        string='Journal Entry',
         index=True,
+        copy=False,
         check_company=True)
     journal_id = fields.Many2one(
         comodel_name='account.journal',
-        inherited=True,
-        related='move_id.journal_id', store=True, readonly=False, precompute=True,
+        compute='_compute_journal_id', store=True, readonly=False, precompute=True,
+        check_company=True,
         index=False,  # covered by account_payment_journal_id_company_id_idx
         required=True,
     )
     company_id = fields.Many2one(
         comodel_name='res.company',
-        inherited=True,
-        related='move_id.company_id', store=True, readonly=False, precompute=True,
+        compute='_compute_company_id', store=True, readonly=False, precompute=True,
         index=False,  # covered by account_payment_journal_id_company_id_idx
         required=True
     )
+    state = fields.Selection(
+        selection=[
+            ('draft', "Draft"),
+            ('in_process', "In Process"),
+            ('paid', "Paid"),
+            ('canceled', "Canceled"),
+            ('rejected', "Rejected"),
+        ],
+        required=True,
+        default='draft',
+        compute='_compute_state', store=True, readonly=False,
+        copy=False,
+    )
     is_reconciled = fields.Boolean(string="Is Reconciled", store=True,
         compute='_compute_reconciliation_status')
     is_matched = fields.Boolean(string="Is Matched With a Bank Statement", store=True,
         compute='_compute_reconciliation_status')
+    is_sent = fields.Boolean(string="Is Sent", readonly=True, copy=False)
     available_partner_bank_ids = fields.Many2many(
         comodel_name='res.partner.bank',
         compute='_compute_available_partner_bank_ids',
@@ -49,10 +64,6 @@ class AccountPayment(models.Model):
         check_company=True,
         ondelete='restrict',
     )
-    is_internal_transfer = fields.Boolean(string="Internal Transfer",
-        readonly=False, store=True,
-        tracking=True,
-        compute="_compute_is_internal_transfer")
     qr_code = fields.Html(string="QR Code URL",
         compute="_compute_qr_code")
     paired_internal_transfer_payment_id = fields.Many2one('account.payment',
@@ -84,7 +95,6 @@ class AccountPayment(models.Model):
         compute='_compute_available_journal_ids'
     )
 
-    # == Synchronized fields with the account.move.lines ==
     amount = fields.Monetary(currency_field='currency_id')
     payment_type = fields.Selection([
         ('outbound', 'Send'),
@@ -94,6 +104,7 @@ class AccountPayment(models.Model):
         ('customer', 'Customer'),
         ('supplier', 'Vendor'),
     ], default='customer', tracking=True, required=True)
+    memo = fields.Char(string="Memo", tracking=True)
     payment_reference = fields.Char(string="Payment Reference", copy=False, tracking=True,
         help="Reference of the document used to issue this payment. Eg. check number, file name, etc.")
     currency_id = fields.Many2one(
@@ -101,6 +112,7 @@ class AccountPayment(models.Model):
         string='Currency',
         compute='_compute_currency_id', store=True, readonly=False, precompute=True,
         help="The payment's currency.")
+    company_currency_id = fields.Many2one(string="Company Currency", related='company_id.currency_id')
     partner_id = fields.Many2one(
         comodel_name='res.partner',
         string="Customer/Vendor",
@@ -124,14 +136,15 @@ class AccountPayment(models.Model):
         compute='_compute_destination_account_id',
         domain="[('account_type', 'in', ('asset_receivable', 'liability_payable'))]",
         check_company=True)
-    destination_journal_id = fields.Many2one(
-        comodel_name='account.journal',
-        string='Destination Journal',
-        domain="[('type', 'in', ('bank', 'cash', 'credit')), ('id', '!=', journal_id)]",
-        check_company=True,
-    )
 
     # == Stat buttons ==
+    invoice_ids = fields.Many2many(  # contains the invoice even if they don't have a journal entry and are not reconciled
+        string="Invoices",
+        comodel_name='account.move',
+        relation='account_move__account_payment',
+        column1='payment_id',
+        column2='invoice_id',
+    )
     reconciled_invoice_ids = fields.Many2many('account.move', string="Reconciled Invoices",
         compute='_compute_stat_buttons_from_reconciliation',
         help="Invoices whose journal items have been reconciled with these payments.")
@@ -165,6 +178,7 @@ class AccountPayment(models.Model):
         compute='_compute_payment_receipt_title'
     )
 
+    need_cancel_request = fields.Boolean(related='move_id.need_cancel_request')
     # used to know whether the field `partner_bank_id` needs to be displayed or not in the payments form views
     show_partner_bank_account = fields.Boolean(
         compute='_compute_show_require_partner_bank')
@@ -178,7 +192,8 @@ class AccountPayment(models.Model):
     amount_company_currency_signed = fields.Monetary(
         currency_field='company_currency_id', compute='_compute_amount_company_currency_signed', store=True)
     # used to get and display duplicate move warning if partner, amount and date match existing payments
-    duplicate_move_ids = fields.Many2many(comodel_name='account.move', compute='_compute_duplicate_move_ids')
+    duplicate_payment_ids = fields.Many2many(comodel_name='account.payment', compute='_compute_duplicate_payment_ids')
+    attachment_ids = fields.One2many('ir.attachment', 'res_id', string='Attachments')
 
     _sql_constraints = [
         (
@@ -243,13 +258,9 @@ class AccountPayment(models.Model):
         return lines
 
     def _get_valid_liquidity_accounts(self):
-        journal_comp = self.journal_id.company_id or self.env.company
-        accessible_branches = journal_comp.with_company(journal_comp)._accessible_branches()
         return (
             self.journal_id.default_account_id |
             self.payment_method_line_id.payment_account_id |
-            accessible_branches.account_journal_payment_debit_account_id |
-            accessible_branches.account_journal_payment_credit_account_id |
             self.journal_id.inbound_payment_method_line_ids.payment_account_id |
             self.journal_id.outbound_payment_method_line_ids.payment_account_id
         )
@@ -265,16 +276,13 @@ class AccountPayment(models.Model):
             ]
         """
         self.ensure_one()
-        if self.is_internal_transfer:
-            label = _("Internal Transfer")
-        else:
-            label = self.payment_method_line_id.name if self.payment_method_line_id else _("No Payment Method")
+        label = self.payment_method_line_id.name if self.payment_method_line_id else _("No Payment Method")
 
-        if self.ref:
+        if self.memo:
             return [
                 ('label', label),
                 ('sep', ": "),
-                ('memo', self.ref),
+                ('memo', self.memo),
             ]
         return [
             ('label', label),
@@ -290,7 +298,7 @@ class AccountPayment(models.Model):
         :return: A list of python dictionary to be passed to the account.move.line's 'create' method.
         '''
         self.ensure_one()
-        write_off_line_vals = write_off_line_vals or {}
+        write_off_line_vals = write_off_line_vals or []
 
         if not self.outstanding_account_id:
             raise UserError(_(
@@ -359,6 +367,45 @@ class AccountPayment(models.Model):
     # COMPUTE METHODS
     # -------------------------------------------------------------------------
 
+    @api.depends('move_id.name', 'state')
+    def _compute_name(self):
+        for payment in self:
+            if payment.id and not payment.name and payment.state in ('in_process', 'paid'):
+                payment.name = (
+                    payment.move_id.name
+                    or self.env['ir.sequence'].with_company(payment.company_id).next_by_code(
+                        'account.payment',
+                        sequence_date=payment.date,
+                    )
+                )
+
+    @api.depends('company_id')
+    def _compute_journal_id(self):
+        for payment in self:
+            company = self.company_id or self.env.company
+            payment.journal_id = self.env['account.journal'].search([
+                *self.env['account.journal']._check_company_domain(company),
+                ('type', 'in', ['bank', 'cash', 'credit']),
+            ], limit=1)
+
+    @api.depends('journal_id')
+    def _compute_company_id(self):
+        for payment in self:
+            if payment.journal_id.company_id not in payment.company_id.parent_ids:
+                payment.company_id = (payment.journal_id.company_id or self.env.company)._accessible_branches()[:1]
+
+    @api.depends('invoice_ids.payment_state')
+    def _compute_state(self):
+        for payment in self:
+            if not payment.state:
+                payment.state = 'draft'
+            if (
+                not payment.move_id
+                and payment.invoice_ids
+                and all(invoice.payment_state == 'paid' for invoice in payment.invoice_ids)
+            ):
+                payment.state = 'paid'
+
     @api.depends('move_id.line_ids.amount_residual', 'move_id.line_ids.amount_residual_currency', 'move_id.line_ids.account_id')
     def _compute_reconciliation_status(self):
         ''' Compute the field indicating if the payments are already reconciled with something.
@@ -368,7 +415,7 @@ class AccountPayment(models.Model):
         for pay in self:
             liquidity_lines, counterpart_lines, writeoff_lines = pay._seek_for_lines()
 
-            if not pay.currency_id or not pay.id:
+            if not pay.currency_id or not pay.id or not pay.move_id:
                 pay.is_reconciled = False
                 pay.is_matched = False
             elif pay.currency_id.is_zero(pay.amount):
@@ -405,11 +452,19 @@ class AccountPayment(models.Model):
                 payment.show_partner_bank_account = payment.payment_method_code in self._get_method_codes_using_bank_account()
             payment.require_partner_bank_account = payment.state == 'draft' and payment.payment_method_code in self._get_method_codes_needing_bank_account()
 
-    @api.depends('amount_total_signed', 'payment_type')
+    @api.depends('move_id.amount_total_signed', 'amount', 'payment_type', 'currency_id', 'date', 'company_id', 'company_currency_id')
     def _compute_amount_company_currency_signed(self):
         for payment in self:
-            liquidity_lines = payment._seek_for_lines()[0]
-            payment.amount_company_currency_signed = sum(liquidity_lines.mapped('balance'))
+            if payment.move_id:
+                liquidity_lines = payment._seek_for_lines()[0]
+                payment.amount_company_currency_signed = sum(liquidity_lines.mapped('balance'))
+            else:
+                payment.amount_company_currency_signed = payment.currency_id._convert(
+                    from_amount=payment.amount,
+                    to_currency=payment.company_currency_id,
+                    company=payment.company_id,
+                    date=payment.date,
+                )
 
     @api.depends('amount', 'payment_type')
     def _compute_amount_signed(self):
@@ -419,13 +474,11 @@ class AccountPayment(models.Model):
             else:
                 payment.amount_signed = payment.amount
 
-    @api.depends('partner_id', 'company_id', 'payment_type', 'destination_journal_id', 'is_internal_transfer')
+    @api.depends('partner_id', 'company_id', 'payment_type')
     def _compute_available_partner_bank_ids(self):
         for pay in self:
             if pay.payment_type == 'inbound':
                 pay.available_partner_bank_ids = pay.journal_id.bank_account_id
-            elif pay.is_internal_transfer:
-                pay.available_partner_bank_ids = pay.destination_journal_id.bank_account_id
             else:
                 pay.available_partner_bank_ids = pay.partner_id.bank_ids\
                         .filtered(lambda x: x.company_id.id in (False, pay.company_id.id))._origin
@@ -437,13 +490,6 @@ class AccountPayment(models.Model):
             if pay.partner_bank_id not in pay.available_partner_bank_ids:
                 pay.partner_bank_id = pay.available_partner_bank_ids[:1]._origin
 
-    @api.depends('partner_id', 'journal_id', 'destination_journal_id')
-    def _compute_is_internal_transfer(self):
-        for payment in self:
-            payment.is_internal_transfer = payment.partner_id \
-                                           and payment.partner_id == payment.journal_id.company_id.partner_id \
-                                           and payment.destination_journal_id
-
     @api.depends('available_payment_method_line_ids')
     def _compute_payment_method_line_id(self):
         ''' Compute the 'payment_method_line_id' field.
@@ -499,35 +545,24 @@ class AccountPayment(models.Model):
         for pay in self:
             pay.currency_id = pay.journal_id.currency_id or pay.journal_id.company_id.currency_id
 
-    @api.depends('is_internal_transfer')
+    @api.depends('journal_id')
     def _compute_partner_id(self):
         for pay in self:
-            if pay.is_internal_transfer:
-                pay.partner_id = pay.journal_id.company_id.partner_id
-            elif pay.partner_id == pay.journal_id.company_id.partner_id:
+            if pay.partner_id == pay.journal_id.company_id.partner_id:
                 pay.partner_id = False
             else:
                 pay.partner_id = pay.partner_id
 
-    @api.depends('journal_id', 'payment_type', 'payment_method_line_id')
+    @api.depends('payment_method_line_id')
     def _compute_outstanding_account_id(self):
         for pay in self:
-            if pay.payment_type == 'inbound':
-                pay.outstanding_account_id = (pay.payment_method_line_id.payment_account_id
-                                              or pay.journal_id.company_id.account_journal_payment_debit_account_id)
-            elif pay.payment_type == 'outbound':
-                pay.outstanding_account_id = (pay.payment_method_line_id.payment_account_id
-                                              or pay.journal_id.company_id.account_journal_payment_credit_account_id)
-            else:
-                pay.outstanding_account_id = False
+            pay.outstanding_account_id = pay.payment_method_line_id.payment_account_id
 
-    @api.depends('journal_id', 'partner_id', 'partner_type', 'is_internal_transfer', 'destination_journal_id')
+    @api.depends('journal_id', 'partner_id', 'partner_type')
     def _compute_destination_account_id(self):
         self.destination_account_id = False
         for pay in self:
-            if pay.is_internal_transfer:
-                pay.destination_account_id = pay.destination_journal_id.company_id.transfer_account_id
-            elif pay.partner_type == 'customer':
+            if pay.partner_type == 'customer':
                 # Receive money from invoice or send money to refund it.
                 if pay.partner_id:
                     pay.destination_account_id = pay.partner_id.with_company(pay.company_id).property_account_receivable_id
@@ -548,11 +583,11 @@ class AccountPayment(models.Model):
                         ('deprecated', '=', False),
                     ], limit=1)
 
-    @api.depends('partner_bank_id', 'amount', 'ref', 'currency_id', 'journal_id', 'move_id.state',
+    @api.depends('partner_bank_id', 'amount', 'memo', 'currency_id', 'journal_id', 'move_id.state',
                  'payment_method_line_id', 'payment_type')
     def _compute_qr_code(self):
         for pay in self:
-            if pay.state in ('draft', 'posted') \
+            if pay.state in ('draft', 'in_process') \
                 and pay.partner_bank_id \
                 and pay.partner_bank_id.allow_out_payment \
                 and pay.payment_method_line_id.code == 'manual' \
@@ -560,7 +595,7 @@ class AccountPayment(models.Model):
                 and pay.currency_id:
 
                 if pay.partner_bank_id:
-                    qr_code = pay.partner_bank_id.build_qr_code_base64(pay.amount, pay.ref, pay.ref, pay.currency_id, pay.partner_id)
+                    qr_code = pay.partner_bank_id.build_qr_code_base64(pay.amount, pay.memo, pay.memo, pay.currency_id, pay.partner_id)
                 else:
                     qr_code = None
 
@@ -583,7 +618,7 @@ class AccountPayment(models.Model):
         if not stored_payments:
             self.reconciled_invoice_ids = False
             self.reconciled_invoices_count = 0
-            self.reconciled_invoices_type = ''
+            self.reconciled_invoices_type = False
             self.reconciled_bill_ids = False
             self.reconciled_bills_count = 0
             self.reconciled_statement_line_ids = False
@@ -591,7 +626,7 @@ class AccountPayment(models.Model):
             return
 
         self.env['account.payment'].flush_model(fnames=['move_id', 'outstanding_account_id'])
-        self.env['account.move'].flush_model(fnames=['move_type', 'payment_id', 'statement_line_id'])
+        self.env['account.move'].flush_model(fnames=['move_type', 'origin_payment_id', 'statement_line_id'])
         self.env['account.move.line'].flush_model(fnames=['move_id', 'account_id', 'statement_line_id'])
         self.env['account.partial.reconcile'].flush_model(fnames=['debit_move_id', 'credit_move_id'])
 
@@ -622,16 +657,21 @@ class AccountPayment(models.Model):
             'payment_ids': tuple(stored_payments.ids)
         })
         query_res = self._cr.dictfetchall()
-        self.reconciled_invoice_ids = self.reconciled_invoices_count = False
-        self.reconciled_bill_ids = self.reconciled_bills_count = False
+
+        for pay in self:
+            pay.reconciled_invoice_ids = pay.invoice_ids.filtered(lambda m: m.is_sale_document(True))
+            pay.reconciled_bill_ids = pay.invoice_ids.filtered(lambda m: m.is_purchase_document(True))
+
         for res in query_res:
             pay = self.browse(res['id'])
             if res['move_type'] in self.env['account.move'].get_sale_types(True):
-                pay.reconciled_invoice_ids += self.env['account.move'].browse(res.get('invoice_ids', []))
-                pay.reconciled_invoices_count = len(res.get('invoice_ids', []))
+                pay.reconciled_invoice_ids |= self.env['account.move'].browse(res.get('invoice_ids', []))
             else:
-                pay.reconciled_bill_ids += self.env['account.move'].browse(res.get('invoice_ids', []))
-                pay.reconciled_bills_count = len(res.get('invoice_ids', []))
+                pay.reconciled_bill_ids |= self.env['account.move'].browse(res.get('invoice_ids', []))
+
+        for pay in self:
+            pay.reconciled_invoices_count = len(pay.reconciled_invoice_ids)
+            pay.reconciled_bills_count = len(pay.reconciled_bill_ids)
 
         self._cr.execute('''
             SELECT
@@ -673,14 +713,14 @@ class AccountPayment(models.Model):
         self.payment_receipt_title = _('Payment Receipt')
 
     @api.depends('partner_id', 'amount', 'date', 'payment_type')
-    def _compute_duplicate_move_ids(self):
+    def _compute_duplicate_payment_ids(self):
         """ Retrieve move ids with same partner_id, amount and date as the current payment """
         payment_to_duplicate_move = self._fetch_duplicate_reference()
         for payment in self:
             # Uses payment._origin.id to handle records in edition/existing records and 0 for new records
-            payment.duplicate_move_ids = payment_to_duplicate_move.get(payment._origin.id, self.env['account.move'])
+            payment.duplicate_payment_ids = payment_to_duplicate_move.get(payment._origin.id, self.env['account.payment'])
 
-    def _fetch_duplicate_reference(self, matching_states=('draft', 'posted')):
+    def _fetch_duplicate_reference(self, matching_states=('draft', 'in_process')):
         """ Retrieve move ids for possible duplicates of payments. Duplicates moves:
         - Have the same partner_id, amount and date as the payment
         - Are not reconciled
@@ -689,110 +729,54 @@ class AccountPayment(models.Model):
          outstanding counterpart can be matched, or
         - Are in the suspense account
         """
-        def build_query(move_table_and_alias, outstanding_account_ids, payments):
-            suspense_account_ids = all_journals.suspense_account_id.ids
-            extra_conditions = SQL(
-                """
-                AND (
-                   -- Case 1: a move is a credit in same account receivable or debit in same acc payable as the payment
-                   (dup_move_line.account_id != ANY(%(outstanding_account_ids)s) AND move_line.balance = dup_move_line.balance)
-                   OR
-                   -- Case 2: a move is a credit in outstanding receipts or debit in outstanding payments
-                   (dup_move_line.account_id = ANY(%(outstanding_account_ids)s) AND move_line.balance = -1.0 * dup_move_line.balance)
-               )
-            """,
-                outstanding_account_ids=outstanding_account_ids,
-            ) if outstanding_account_ids else SQL("AND move_line.balance = dup_move_line.balance")
-
-            return SQL(
-                """
-                SELECT
-                       move_line.payment_id,
-                       array_agg(DISTINCT dup_move_line.move_id) AS duplicate_move_ids
-                  FROM %(move_table_and_alias)s
-                  JOIN account_move_line AS dup_move_line
-                    ON move_line.move_id != dup_move_line.move_id
-                   AND move_line.partner_id = dup_move_line.partner_id
-                   AND move_line.company_id = dup_move_line.company_id
-                   AND move_line.date = dup_move_line.date
-                   AND dup_move_line.parent_state = ANY(%(matching_states)s)
-                   AND (
-                       move_line.account_id = dup_move_line.account_id
-                       OR dup_move_line.account_id = ANY(%(account_ids)s)
-                   )
-                   AND NOT dup_move_line.reconciled
-                 WHERE move_line.payment_id = ANY(%(payments)s)
-                   %(extra_conditions)s
-                   AND (
-                       move_line.payment_type = 'inbound' AND dup_move_line.balance < 0.0
-                       OR move_line.payment_type = 'outbound' AND dup_move_line.balance > 0.0
-                   )
-              GROUP BY move_line.payment_id
-            """,
-                move_table_and_alias=move_table_and_alias,
-                matching_states=list(matching_states),
-                account_ids=suspense_account_ids + outstanding_account_ids,
-                payments=payments,
-                extra_conditions=extra_conditions,
-            )
-
         # Does not perform unnecessary check if partner_id or amount are not set, nor if payment is posted
-        if not self.filtered(lambda p: p.partner_id and p.amount and p.state != 'posted'):
+        payments = self.filtered(lambda p: p.partner_id and p.amount and p.state != 'in_process')
+        if not payments:
             return {}
-        # Separate inbound and outbound payments, as their outstanding accounts differ and need to be checked separately
-        payments_inbound = self.filtered(lambda p: p.payment_type == 'inbound')
-        payments_outbound = self.filtered(lambda p: p.payment_type == 'outbound')
-        all_journals = self.env['account.journal'].search([('type', 'in', ('bank', 'cash'))])
-        if payments_inbound:
-            inbound_outstanding_account_ids = all_journals.inbound_payment_method_line_ids.payment_account_id.ids + self.company_id.account_journal_payment_debit_account_id.ids
-        if payments_outbound:
-            outbound_outstanding_account_ids = all_journals.outbound_payment_method_line_ids.payment_account_id.ids + self.company_id.account_journal_payment_credit_account_id.ids
 
         # Update tables involved in the query
-        self.env['account.move.line'].flush_model(('move_id', 'payment_id', 'balance', 'account_id', 'company_id', 'date', 'partner_id'))
-        self.env['account.payment'].flush_model(('move_id', 'payment_type'))
+        used_fields = ("company_id", "partner_id", "date", "state", "amount", 'payment_type')
+        self.flush_model(used_fields)
 
+        payment_table_and_alias = SQL("account_payment AS payment")
         if not self[0].id:  # if record is under creation/edition in UI, safely inject values in the query
             # Necessary since new record aren't searchable in the DB and record in edition aren't up to date yet
-            place_holders = {
-                'move_id': self._origin.move_id.id or 0,
-                'payment_id': self._origin.id or 0,
-                'payment_type': self.payment_type,
-                'balance': self.amount if self.payment_type == 'outbound' else -self.amount,
-                'account_id': self.destination_account_id.id,
-                'company_id': self.company_id.id or None,
-                'date': self.date or None,
-                'partner_id': self.partner_id.id,
+            values = {
+                field_name: self._fields[field_name].convert_to_write(self[field_name], self) or None
+                for field_name in used_fields
             }
-            # In case of null values, postgres may have issues with the column type, so we cast the columns
-            move_table_and_alias = SQL("""
-                (VALUES (%(move_id)s::int4, %(payment_id)s::int4, %(payment_type)s::varchar, %(balance)s::numeric,%(account_id)s::int4, %(company_id)s::int4, %(date)s::date, %(partner_id)s::int4))
-                AS move_line(move_id, payment_id, payment_type, balance, account_id, company_id, date, partner_id)
-            """, **place_holders)
-            outstanding_account_ids = inbound_outstanding_account_ids if self.payment_type == 'inbound' else outbound_outstanding_account_ids
-            query = build_query(move_table_and_alias, outstanding_account_ids, [place_holders['payment_id']])
-
-        else:
-            move_table_and_alias = SQL("""
-                (SELECT account_move_line.*, account_payment.payment_type
-                   FROM account_move_line
-                   JOIN account_payment
-                     ON account_payment.move_id = account_move_line.move_id) AS move_line
-            """)
-            if payments_inbound and payments_outbound:
-                query_inbound = build_query(move_table_and_alias, inbound_outstanding_account_ids, payments_inbound.ids)
-                query_outbound = build_query(move_table_and_alias, outbound_outstanding_account_ids, payments_outbound.ids)
-                query = SQL(
-                    "%(query_inbound)s UNION %(query_outbound)s",
-                    query_inbound=query_inbound,
-                    query_outbound=query_outbound,
-                )
-            else:
-                outstanding_account_ids = inbound_outstanding_account_ids if payments_inbound else outbound_outstanding_account_ids
-                query = build_query(move_table_and_alias, outstanding_account_ids, self.ids)
+            values["id"] = self._origin.id or 0
+            # The amount total depends on the field line_ids and is calculated upon saving, we needed a way to get it even when the
+            # invoices has not been saved yet.
+            casted_values = SQL(', ').join(
+                SQL("%s::%s", value, SQL.identifier(self._fields[field_name].column_type[0]))
+                for field_name, value in values.items()
+            )
+            column_names = SQL(', ').join(SQL.identifier(field_name) for field_name in values)
+            payment_table_and_alias = SQL("(VALUES (%s)) AS payment(%s)", casted_values, column_names)
+
+        query = SQL(
+            """
+                SELECT payment.id AS payment_id,
+                       ARRAY_AGG(DISTINCT duplicate_payment.id) AS duplicate_payment_ids
+                  FROM %(payment_table_and_alias)s
+                  JOIN account_payment AS duplicate_payment ON payment.id != duplicate_payment.id
+                                                           AND payment.partner_id = duplicate_payment.partner_id
+                                                           AND payment.company_id = duplicate_payment.company_id
+                                                           AND payment.date = duplicate_payment.date
+                                                           AND payment.payment_type = duplicate_payment.payment_type
+                                                           AND payment.amount = duplicate_payment.amount
+                                                           AND duplicate_payment.state IN %(matching_states)s
+                 WHERE payment.id = ANY(%(payments)s)
+              GROUP BY payment.id
+            """,
+            payment_table_and_alias=payment_table_and_alias,
+            matching_states=tuple(matching_states),
+            payments=payments.ids or [0],
+        )
 
         return {
-            payment_id: self.env['account.move'].browse(duplicate_ids)
+            payment_id: self.env['account.payment'].browse(duplicate_ids)
             for payment_id, duplicate_ids in self.env.execute_query(query)
         }
 
@@ -800,13 +784,6 @@ class AccountPayment(models.Model):
     # ONCHANGE METHODS
     # -------------------------------------------------------------------------
 
-    @api.onchange('posted_before', 'state', 'journal_id', 'date')
-    def _onchange_journal_date(self):
-        # Before the record is created, the move_id doesn't exist yet, and the name will not be
-        # recomputed correctly if we change the journal or the date, leading to inconsitencies
-        if not self.move_id:
-            self.name = False
-
     @api.onchange('partner_id')
     def _inverse_partner_id(self):
         """
@@ -841,25 +818,27 @@ class AccountPayment(models.Model):
             elif pay.payment_method_line_id.journal_id and pay.payment_method_line_id.journal_id != pay.journal_id:
                 raise ValidationError(_("The selected payment method is not available for this payment, please select the payment method again."))
 
+    @api.constrains('state', 'move_id')
+    def _check_move_id(self):
+        for payment in self:
+            if (
+                payment.state not in ('draft', 'canceled')
+                and not payment.move_id
+                and payment.outstanding_account_id
+            ):
+                raise ValidationError(_("A payment with an outstanding account cannot be confirmed without having a journal entry."))
+            if payment.state == 'draft' and payment.move_id:
+                raise ValidationError(_("A payment cannot have a journal entry if it is not confirmed."))
     # -------------------------------------------------------------------------
     # LOW-LEVEL METHODS
     # -------------------------------------------------------------------------
 
-    def default_get(self, fields_list):
-        self_ctx = self.with_context(is_payment=True)
-        defaults = super(AccountPayment, self_ctx).default_get(fields_list)
-        if 'journal_id' in fields_list and not defaults.get('journal_id'):
-            defaults['journal_id'] = self_ctx.env['account.move']._search_default_journal().id
-        return defaults
-
-    def new(self, values=None, origin=None, ref=None):
-        return super(AccountPayment, self.with_context(is_payment=True)).new(values, origin, ref)
-
     @api.model_create_multi
     def create(self, vals_list):
         # OVERRIDE
         write_off_line_vals_list = []
         force_balance_vals_list = []
+        linecomplete_line_vals_list = []
 
         for vals in vals_list:
 
@@ -869,240 +848,86 @@ class AccountPayment(models.Model):
             # Hack to force a custom balance.
             force_balance_vals_list.append(vals.pop('force_balance', None))
 
-            # Force the move_type to avoid inconsistency with residual 'default_move_type' inside the context.
-            vals['move_type'] = 'entry'
+            # Hack to add a custom line.
+            linecomplete_line_vals_list.append(vals.pop('line_ids', None))
 
-        payments = super(AccountPayment, self.with_context(is_payment=True)).create([{
-            'name': False,
-            **vals,
-        } for vals in vals_list])
+        payments = super().create(vals_list)
 
         for i, (pay, vals) in enumerate(zip(payments, vals_list)):
-            # Write payment_id on the journal entry plus the fields being stored in both models but having the same
-            # name, e.g. partner_bank_id. The ORM is currently not able to perform such synchronization and make things
-            # more difficult by creating related fields on the fly to handle the _inherits.
-            # Then, when partner_bank_id is in vals, the key is consumed by account.payment but is never written on
-            # account.move.
-            to_write = {'payment_id': pay.id, 'name': False}
-            for k, v in vals_list[i].items():
-                if k in self._fields and self._fields[k].store and k in pay.move_id._fields and pay.move_id._fields[k].store:
-                    to_write[k] = v
-
-            if 'line_ids' not in vals_list[i]:
-                to_write['line_ids'] = [
-                    Command.create(line_vals)
-                    for line_vals in pay._prepare_move_line_default_vals(
-                        write_off_line_vals=write_off_line_vals_list[i],
-                        force_balance=force_balance_vals_list[i],
-                    )
-                ]
-            with self.env.protecting(self.env['account.move']._get_protected_vals(vals, pay)):
-                pay.move_id.write(to_write)
-            self.env.add_to_compute(self.env['account.move']._fields['name'], pay.move_id)
-
-        # We need to reset the cached name, since it was recomputed on the delegate account.move model
-        payments.invalidate_recordset(fnames=['name'])
-        return payments.with_env(self.env)  # clear the context
+            if (
+                write_off_line_vals_list[i] is not None
+                or force_balance_vals_list[i] is not None
+                or linecomplete_line_vals_list[i] is not None
+            ):
+                pay._generate_journal_entry(
+                    write_off_line_vals=write_off_line_vals_list[i],
+                    force_balance=force_balance_vals_list[i],
+                    line_ids=linecomplete_line_vals_list[i],
+                )
+                # propagate the related fields to the move as it is being created after the payment
+                if move_vals := {
+                    fname: value
+                    for fname, value in vals.items()
+                    if self._fields[fname].related and (self._fields[fname].related or '').split('.')[0] == 'move_id'
+                }:
+                    pay.move_id.write(move_vals)
+        return payments
 
     def write(self, vals):
-        # OVERRIDE
-        res = super().write(vals)
-        self._synchronize_to_moves(set(vals.keys()))
-        return res
+        if vals.get('state') == 'in_process' and not vals.get('move_id'):
+            self.filtered(lambda p: not p.move_id)._generate_journal_entry()
+            self.move_id.action_post()
+        return super().write(vals)
 
     def unlink(self):
-        # OVERRIDE to unlink the inherited account.move (move_id field) as well.
-        moves = self.with_context(force_delete=True).move_id
+        linked_invoices = self.invoice_ids
         res = super().unlink()
-        moves.unlink()
+        self.env.add_to_compute(linked_invoices._fields['payment_state'], linked_invoices)
         return res
 
     @api.depends('move_id.name')
     def _compute_display_name(self):
         for payment in self:
-            payment.display_name = payment.move_id.name if payment.move_id.name != '/' else _('Draft Payment')
+            payment.display_name = payment.name or _('Draft Payment')
 
     def copy(self, default=None):
-        if not self.is_internal_transfer:
-            default = {
-                'journal_id': self.journal_id.id,
-                'payment_method_line_id': self.payment_method_line_id.id,
-                **(default or {}),
-            }
-        return super().copy(default=default)
+        return super().copy(default={
+            'journal_id': self.journal_id.id,
+            'payment_method_line_id': self.payment_method_line_id.id,
+            **(default or {}),
+        })
 
     # -------------------------------------------------------------------------
     # SYNCHRONIZATION account.payment <-> account.move
     # -------------------------------------------------------------------------
 
-    def _synchronize_from_moves(self, changed_fields):
-        ''' Update the account.payment regarding its related account.move.
-        Also, check both models are still consistent.
-        :param changed_fields: A set containing all modified fields on account.move.
-        '''
-        if self._context.get('skip_account_move_synchronization'):
-            return
-
-        for pay in self.with_context(skip_account_move_synchronization=True):
-
-            # After the migration to 14.0, the journal entry could be shared between the account.payment and the
-            # account.bank.statement.line. In that case, the synchronization will only be made with the statement line.
-            if pay.move_id.statement_line_id:
-                continue
-
-            move = pay.move_id
-            move_vals_to_write = {}
-            payment_vals_to_write = {}
-
-            if 'journal_id' in changed_fields:
-                if pay.journal_id.type not in ('bank', 'cash', 'credit'):
-                    raise UserError(_("A payment must always belongs to a bank, cash or credit journal."))
-
-            if 'line_ids' in changed_fields:
-                all_lines = move.line_ids
-                liquidity_lines, counterpart_lines, writeoff_lines = pay._seek_for_lines()
-
-                if len(liquidity_lines) != 1:
-                    raise UserError(_(
-                        "Journal Entry %s is not valid. In order to proceed, the journal items must "
-                        "include one and only one outstanding payments/receipts account.",
-                        move.display_name,
-                    ))
-
-                if len(counterpart_lines) != 1:
-                    raise UserError(_(
-                        "Journal Entry %s is not valid. In order to proceed, the journal items must "
-                        "include one and only one receivable/payable account (with an exception of "
-                        "internal transfers).",
-                        move.display_name,
-                    ))
-
-                if any(line.currency_id != all_lines[0].currency_id for line in all_lines):
-                    raise UserError(_(
-                        "Journal Entry %s is not valid. In order to proceed, the journal items must "
-                        "share the same currency.",
-                        move.display_name,
-                    ))
-
-                if any(line.partner_id != all_lines[0].partner_id for line in all_lines):
-                    raise UserError(_(
-                        "Journal Entry %s is not valid. In order to proceed, the journal items must "
-                        "share the same partner.",
-                        move.display_name,
-                    ))
-
-                if counterpart_lines.account_id.account_type == 'asset_receivable':
-                    partner_type = 'customer'
-                else:
-                    partner_type = 'supplier'
-
-                liquidity_amount = liquidity_lines.amount_currency
-
-                move_vals_to_write.update({
-                    'currency_id': liquidity_lines.currency_id.id,
-                    'partner_id': liquidity_lines.partner_id.id,
-                })
-                payment_vals_to_write.update({
-                    'amount': abs(liquidity_amount),
-                    'partner_type': partner_type,
-                    'currency_id': liquidity_lines.currency_id.id,
-                    'destination_account_id': counterpart_lines.account_id.id,
-                    'partner_id': liquidity_lines.partner_id.id,
-                })
-                if liquidity_amount > 0.0:
-                    payment_vals_to_write.update({'payment_type': 'inbound'})
-                elif liquidity_amount < 0.0:
-                    payment_vals_to_write.update({'payment_type': 'outbound'})
-
-            move.write(move._cleanup_write_orm_values(move, move_vals_to_write))
-            pay.write(move._cleanup_write_orm_values(pay, payment_vals_to_write))
-
-    @api.model
-    def _get_trigger_fields_to_synchronize(self):
-        return (
-            'date', 'amount', 'payment_type', 'partner_type', 'payment_reference', 'is_internal_transfer',
-            'currency_id', 'partner_id', 'destination_account_id', 'partner_bank_id', 'journal_id'
-        )
-
-    def _synchronize_to_moves(self, changed_fields):
-        ''' Update the account.move regarding the modified account.payment.
-        :param changed_fields: A list containing all modified fields on account.payment.
-        '''
-        if self._context.get('skip_account_move_synchronization'):
-            return
-
-        if not any(field_name in changed_fields for field_name in self._get_trigger_fields_to_synchronize()):
-            return
-
-        for pay in self.with_context(skip_account_move_synchronization=True):
-            liquidity_lines, counterpart_lines, writeoff_lines = pay._seek_for_lines()
-
-            # Make sure to preserve the write-off amount.
-            # This allows to create a new payment with custom 'line_ids'.
-
-            write_off_line_vals = []
-            if liquidity_lines and counterpart_lines and writeoff_lines:
-                write_off_line_vals.append({
-                    'name': writeoff_lines[0].name,
-                    'account_id': writeoff_lines[0].account_id.id,
-                    'partner_id': writeoff_lines[0].partner_id.id,
-                    'currency_id': writeoff_lines[0].currency_id.id,
-                    'amount_currency': sum(writeoff_lines.mapped('amount_currency')),
-                    'balance': sum(writeoff_lines.mapped('balance')),
-                })
-
-            line_vals_list = pay._prepare_move_line_default_vals(write_off_line_vals=write_off_line_vals)
-
-            line_ids_commands = [
-                Command.update(liquidity_lines.id, line_vals_list[0]) if liquidity_lines else Command.create(line_vals_list[0]),
-                Command.update(counterpart_lines.id, line_vals_list[1]) if counterpart_lines else Command.create(line_vals_list[1])
-            ]
-
-            for line in writeoff_lines:
-                line_ids_commands.append((2, line.id))
-
-            for extra_line_vals in line_vals_list[2:]:
-                line_ids_commands.append((0, 0, extra_line_vals))
-
-            # Update the existing journal items.
-            # If dealing with multiple write-off lines, they are dropped and a new one is generated.
-
-            pay.move_id\
-                .with_context(skip_invoice_sync=True)\
-                .write({
-                    'partner_id': pay.partner_id.id,
-                    'currency_id': pay.currency_id.id,
-                    'partner_bank_id': pay.partner_bank_id.id,
-                    'line_ids': line_ids_commands,
-                })
-
-    def _create_paired_internal_transfer_payment(self):
-        ''' When an internal transfer is posted, a paired payment is created
-        with opposite payment_type and swapped journal_id & destination_journal_id.
-        Both payments liquidity transfer lines are then reconciled.
-        '''
-        for payment in self:
-
-            paired_payment = payment.copy({
-                'journal_id': payment.destination_journal_id.id,
-                'company_id': payment.company_id.id,
-                'destination_journal_id': payment.journal_id.id,
-                'payment_type': payment.payment_type == 'outbound' and 'inbound' or 'outbound',
-                'move_id': None,
-                'ref': payment.ref,
-                'paired_internal_transfer_payment_id': payment.id,
-                'date': payment.date,
+    def _generate_journal_entry(self, write_off_line_vals=None, force_balance=None, line_ids=None):
+        need_move = self.filtered(lambda p: not p.move_id and p.outstanding_account_id)
+        assert len(self) == 1 or (not write_off_line_vals and not force_balance and not line_ids)
+
+        move_vals = []
+        for pay in need_move:
+            move_vals.append({
+                'move_type': 'entry',
+                'ref': pay.memo,
+                'date': pay.date,
+                'journal_id': pay.journal_id.id,
+                'company_id': pay.company_id.id,
+                'partner_id': pay.partner_id.id,
+                'currency_id': pay.currency_id.id,
+                'partner_bank_id': pay.partner_bank_id.id,
+                'line_ids': line_ids or [
+                    Command.create(line_vals)
+                    for line_vals in pay._prepare_move_line_default_vals(
+                        write_off_line_vals=write_off_line_vals,
+                        force_balance=force_balance,
+                    )
+                ],
+                'origin_payment_id': pay.id,
             })
-            paired_payment.move_id._post(soft=False)
-            payment.paired_internal_transfer_payment_id = paired_payment
-            body = _("This payment has been created from: %(link)s", link=payment._get_html_link())
-            paired_payment.message_post(body=body)
-            body = _("A second payment has been created: %(link)s", link=paired_payment._get_html_link())
-            payment.message_post(body=body)
-
-            lines = (payment.move_id.line_ids + paired_payment.move_id.line_ids).filtered(
-                lambda l: l.account_id == payment.destination_account_id and not l.reconciled)
-            lines.reconcile()
+        moves = self.env['account.move'].create(move_vals)
+        for pay, move in zip(need_move, moves):
+            pay.write({'move_id': move.id, 'state': 'in_process'})
 
     def _get_payment_receipt_report_values(self):
         """ Get the extra values when rendering the Payment Receipt PDF report.
@@ -1122,10 +947,10 @@ class AccountPayment(models.Model):
     # -------------------------------------------------------------------------
 
     def mark_as_sent(self):
-        self.write({'is_move_sent': True})
+        self.write({'is_sent': True})
 
     def unmark_as_sent(self):
-        self.write({'is_move_sent': False})
+        self.write({'is_sent': False})
 
     def action_post(self):
         ''' draft -> posted '''
@@ -1139,46 +964,36 @@ class AccountPayment(models.Model):
                     partner=payment.partner_id.display_name,
                 ))
 
-        self.move_id._post(soft=False)
+        self.state = 'in_process'
+
+    def action_validate(self):
+        self.state = 'paid'
 
-        self.filtered(
-            lambda pay: pay.is_internal_transfer and not pay.paired_internal_transfer_payment_id
-        )._create_paired_internal_transfer_payment()
+    def action_reject(self):
+        self.state = 'rejected'
 
     def action_cancel(self):
-        ''' draft -> cancelled '''
-        self.move_id.button_cancel()
+        self.state = 'canceled'
+        draft_moves = self.move_id.filtered(lambda m: m.state == 'draft')
+        draft_moves.unlink()
+        (self.move_id - draft_moves).button_cancel()
 
     def button_request_cancel(self):
         return self.move_id.button_request_cancel()
 
     def action_draft(self):
-        ''' posted -> draft '''
-        self.move_id.button_draft()
+        self.state = 'draft'
 
     def button_open_invoices(self):
         ''' Redirect the user to the invoice(s) paid by this payment.
         :return:    An action on account.move.
         '''
         self.ensure_one()
-
-        action = {
-            'name': _("Paid Invoices"),
-            'type': 'ir.actions.act_window',
-            'res_model': 'account.move',
-            'context': {'create': False},
-        }
-        if len(self.reconciled_invoice_ids) == 1:
-            action.update({
-                'view_mode': 'form',
-                'res_id': self.reconciled_invoice_ids.id,
-            })
-        else:
-            action.update({
-                'view_mode': 'list,form',
-                'domain': [('id', 'in', self.reconciled_invoice_ids.ids)],
-            })
-        return action
+        return (self.invoice_ids | self.reconciled_invoice_ids).with_context(
+            create=False
+        )._get_records_action(
+            name=_("Paid Invoices"),
+        )
 
     def button_open_bills(self):
         ''' Redirect the user to the bill(s) paid by this payment.
@@ -1242,22 +1057,6 @@ class AccountPayment(models.Model):
             'res_id': self.move_id.id,
         }
 
-    def action_open_destination_journal(self):
-        ''' Redirect the user to this destination journal.
-        :return:    An action on account.move.
-        '''
-        self.ensure_one()
-
-        action = {
-            'name': _("Destination journal"),
-            'type': 'ir.actions.act_window',
-            'res_model': 'account.journal',
-            'context': {'create': False},
-            'view_mode': 'form',
-            'target': 'new',
-            'res_id': self.destination_journal_id.id,
-        }
-        return action
 
 # For optimization purpose, creating the reverse relation of m2o in _inherits saves
 # a lot of SQL queries

--- a/addons/account/models/chart_template.py
+++ b/addons/account/models/chart_template.py
@@ -731,20 +731,6 @@ class AccountChartTemplate(models.AbstractModel):
                 'code_digits': code_digits,
                 'account_type': 'asset_current',
             },
-            'account_journal_payment_debit_account_id': {
-                'name': _("Outstanding Receipts"),
-                'prefix': bank_prefix,
-                'code_digits': code_digits,
-                'account_type': 'asset_current',
-                'reconcile': True,
-            },
-            'account_journal_payment_credit_account_id': {
-                'name': _("Outstanding Payments"),
-                'prefix': bank_prefix,
-                'code_digits': code_digits,
-                'account_type': 'asset_current',
-                'reconcile': True,
-            },
             'account_journal_early_pay_discount_loss_account_id': {
                 'name': _("Cash Discount Loss"),
                 'code': '999998',

--- a/addons/account/models/company.py
+++ b/addons/account/models/company.py
@@ -100,8 +100,6 @@ class ResCompany(models.Model):
     default_cash_difference_income_account_id = fields.Many2one('account.account', string="Cash Difference Income", check_company=True)
     default_cash_difference_expense_account_id = fields.Many2one('account.account', string="Cash Difference Expense", check_company=True)
     account_journal_suspense_account_id = fields.Many2one('account.account', string='Journal Suspense Account', check_company=True)
-    account_journal_payment_debit_account_id = fields.Many2one('account.account', string='Journal Outstanding Receipts', check_company=True)
-    account_journal_payment_credit_account_id = fields.Many2one('account.account', string='Journal Outstanding Payments', check_company=True)
     account_journal_early_pay_discount_gain_account_id = fields.Many2one(comodel_name='account.account', string='Cash Discount Write-Off Gain Account', check_company=True)
     account_journal_early_pay_discount_loss_account_id = fields.Many2one(comodel_name='account.account', string='Cash Discount Write-Off Loss Account', check_company=True)
     transfer_account_code_prefix = fields.Char(string='Prefix of the transfer accounts')

--- a/addons/account/models/res_config_settings.py
+++ b/addons/account/models/res_config_settings.py
@@ -68,27 +68,6 @@ class ResConfigSettings(models.TransientModel):
         help='Bank Transactions are posted immediately after import or synchronization. '
              'Their counterparty is the bank suspense account.\n'
              'Reconciliation replaces the latter by the definitive account(s).')
-    account_journal_payment_debit_account_id = fields.Many2one(
-        comodel_name='account.account',
-        string='Outstanding Receipts',
-        readonly=False,
-        check_company=True,
-        related='company_id.account_journal_payment_debit_account_id',
-        domain="[('deprecated', '=', False), ('account_type', '=', 'asset_current')]",
-        help='Incoming payments are posted on an Outstanding Receipts Account. '
-             'In the bank reconciliation widget, they appear as blue lines.\n'
-             'Bank transactions are then reconciled on the Outstanding Receipts Accounts rather than the Receivable '
-             'Account.')
-    account_journal_payment_credit_account_id = fields.Many2one(
-        comodel_name='account.account',
-        string='Outstanding Payments',
-        readonly=False,
-        check_company=True,
-        related='company_id.account_journal_payment_credit_account_id',
-        domain="[('deprecated', '=', False), ('account_type', '=', 'asset_current')]",
-        help='Outgoing Payments are posted on an Outstanding Payments Account. '
-             'In the bank reconciliation widget, they appear as blue lines.\n'
-             'Bank transactions are then reconciled on the Outstanding Payments Account rather the Payable Account.')
     transfer_account_id = fields.Many2one('account.account', string="Internal Transfer",
         related='company_id.transfer_account_id', readonly=False,
         check_company=True,
